src/mbgl/renderer/layers/line_layer_tweaker.cpp

```diff
@@ -38,13 +38,16 @@ static const StringIdentity idTexImageName = stringIndexer().get("u_image");
 
 void LineLayerTweaker::execute(LayerGroupBase& layerGroup, const PaintParameters& parameters) {
     auto& context = parameters.context;
+    const auto zoom = parameters.state.getZoom();
     const auto& evaluated = static_cast<const LineLayerProperties&>(*evaluatedProperties).evaluated;
     const auto& crossfade = static_cast<const LineLayerProperties&>(*evaluatedProperties).crossfade;
 
-    const auto zoom = parameters.state.getZoom();
+    // Each property UBO is updated at most once if new evaluated properties were set
+    bool propUpdateFlags[4] = {propertiesUpdated, propertiesUpdated, propertiesUpdated, propertiesUpdated};
+    propertiesUpdated = false;
 
     const auto getLinePropsBuffer = [&]() {
-        if (!linePropertiesBuffer) {
+        if (!linePropertiesBuffer || propUpdateFlags[0]) {
```

Comment:
Could you assign these indices to names to help clarify? ie `constexpr size_t LinePropUpdateFlags = 0;` (or with an enum),